lib/_stream_writable.js

```diff
@@ -51,11 +51,20 @@ function WritableState(options, stream) {
   // Note: 0 is a valid value, means that we always return false if
   // the entire buffer is not flushed immediately on write()
   var hwm = options.highWaterMark;
-  var defaultHwm = this.objectMode ? 16 : 16 * 1024;
-  this.highWaterMark = (hwm || hwm === 0) ? hwm : defaultHwm;
+  if (hwm == null) {
+    this.highWaterMark = this.objectMode ? 16 : 16 * 1024;
+  } else {
+    if (typeof hwm !== 'number') {
+      throw new TypeError('"highWaterMark" must be a number');
+    }
 
-  // cast to ints.
-  this.highWaterMark = Math.floor(this.highWaterMark);
+    if (hwm < 0) {
+      throw new RangeError('"highWaterMark" must not be negative');
```

Comment:
This is not actually the case for streams. It's a way bigger chunk of work that we need to do, as we would have to support them in `readable-stream` as well.